spec/mailers/application_mailer_spec.rb

```diff
@@ -0,0 +1,175 @@
+require 'rails_helper'
+
+require 'email_spec'
+require 'email_spec/rspec'
+
+#  shared examples for RSpec below
+
+shared_examples 'delivery is OK' do
+
+  # Need to check these expectations within the same it block else .deliveries will not be accurate
+  it 'is delivered (delivery count, delivered id OK)' do
+    expect(email_response.message_id).not_to be_nil
+
+    delivered = ApplicationMailer.deliveries
+    expect(delivered.count).to eq 1
+    expect(email_response[:message_id]).to eq(delivered.first[:message_id])
+  end
+
+end
+
+#----------------------------------------
+
+RSpec.describe ApplicationMailer, type: :mailer do
+
+
+  TEST_TEMPLATE = 'empty_template'
+
+
+  describe '#mailgun_client' do
+
+    let(:mg_client) { subject.mailgun_client }
+
+    it 'is never nil' do
+      expect(mg_client).not_to be_nil
+    end
+
+    it 'is a Mailgun::Client.new' do
+      expect(mg_client).to be_an_instance_of(Mailgun::Client)
+    end
+
+  end
+
+
+  describe '#message_builder' do
+
+    let(:mg_builder) { subject.message_builder }
+
+    it 'is never nil' do
+      expect(mg_builder).not_to be_nil
+    end
+
+    it 'is a Mailgun::MessageBuilder.new' do
+      expect(mg_builder).to be_an_instance_of(Mailgun::MessageBuilder)
+    end
+
+    it "default from address is ENV['SHF_SENDER_EMAIL']" do
+      expect(mg_builder.message[:from]).to eq([ENV['SHF_SENDER_EMAIL']])
+    end
+
+  end
+
+
+  describe '#domain' do
+
+    it 'can be nil (which should then rightly mean an error response from Mailgun)' do
+
+      stub_const('ENV', ENV.to_hash)
+      ENV.delete('MAILGUN_DOMAIN')
+
+      expect(subject.domain).to be_nil
+    end
+
+
+    it 'can be some value (not nil)' do
+      stub_const('ENV', ENV.to_hash.merge('MAILGUN_DOMAIN' => 'blorf.com'))
+      expect(subject.domain).to eq 'blorf.com'
+    end
+
+  end
+
+
+  describe 'content is correct' do
+
+    before(:all) do
+      @test_user = create(:user)
```

Comment:
I'm not a great fan of before(:all) because it introduces dependencies between tests, and although it seems innocuous here (apart from the database entries not being cleaned up) I would prefer just something like:

```
let(:test_user) {build(:user)}
let(:email) {ApplicationMailer.test_email(test_user)}
```
I realise a setup per test could mean a performance hit in some cases, but I've tried it out in this specific case and I can't see a noticeable difference.

It would also remove the need for adding a global before(:all) hook to clean the database (which in my opinion is encouraging unwanted behaviour - but I know this is a pet peeve of mine, so just disregard as needed)

Reply:
I agree that overall it's not a performance issue.

My personal style is to refactor common things as much as I can; this is how I tend do it in RSpec.

But as you note, this requires us to insert some specific DatabaseCleaning to handle anything set up in a before(:all) block.

We can discuss this in a team meeting (or in the related PR #358 or Slack or...) and figure out if we want to have `RSpec before(:all)` blocks or not. And then I'll happily adjust. :-)

Reply:
I fully agree on refactoring common things, and also in RSpec. But to me that would mean moving the common code to before(:each). Moving it to before(:all) changes the behaviour.